tests/unit/test_characterProcessing.py

```diff
@@ -0,0 +1,74 @@
+# A part of NonVisual Desktop Access (NVDA)
+# This file is covered by the GNU General Public License.
+# See the file COPYING for more details.
+# Copyright (C) 2020 NV Access Limited
+
+"""Unit tests for the characterProcessing module.
+"""
+
+import unittest
+import re
+from characterProcessing import SYMLVL_ALL
+from characterProcessing import SpeechSymbolProcessor
+from characterProcessing import processSpeechSymbols as process
+
+
+class TestComplex(unittest.TestCase):
+	"""Test the complex symbols rules.
+	"""
+
+	def _replace_cb(self, name, replacement):
+		"""Return a regexp callback which replaces matches of the given
+		group name with the replacement string, with support for
+		replacement of group references.
+		"""
+		def replace(self, m):
+			if m.lastgroup == name:
+				return SpeechSymbolProcessor._replaceGroups(self, m, replacement)
+			return m.group()
+		return self.replace
+
+	def _replace(self, string, pattern, name, replacement):
+		"""Perform a pattern-based replacement on a string, for one
+		named group, with support for replacement of group references.
+		"""
+		regexp = re.compile(pattern, re.UNICODE)
+		return regexp.sub(self._replace_cb(name, replacement), string)
+
+	def test_group_replacement(self):
+		"""Test that plain text gets properly replaced
+		"""
+		self.assertEqual(self._replace("1", r"(?P<foo>(\d))", "foo", "a"), "a")
+
+	def test_backslash_replacement(self):
+		"""Test that backslashes get properly replaced
+		"""
+		self.assertEqual(self._replace("1", r"(?P<foo>(\d))", "foo", r"\\"), "\\")
+
+	def test_unknown_escape(self):
+		"""Test that a non-supported escaped character raises an error
+		"""
+                with self.assertRaises(LookupError):
+                    self._replace("1", r"(?P<foo>(\d))", "foo", r"\a")
+
+	def test_unterminated_escape(self):
+		"""Test that an escape at the end of string raises an error
+		"""
+                with self.assertRaises(LookupError):
+                    self._replace("1", r"(?P<foo>(\d))", "foo", "\\")
+
+	def test_group_replacements(self):
+		"""Test that group references get properly replaced
+		"""
+		self.assertEqual(self._replace("1.2", r"(?P<foo>(\d)\.(\d))", "foo", r"\2 \1"), "2 1")

```

Comment:
I think this would be clearer if it didn't use '2' and '1', since these match the number in the group. Perhaps something like:
```suggestion
		self.assertEqual(self._replace("foo.BAR", r"(?P<foo>([a-z])\.([A-Z))", "foo", r"\2>\1"), "BAR>foo")

```

Reply:
Ok, done so.